### PR TITLE
Add labels

### PR DESCRIPTION
**Pull request by davgur**

### CODE REVIEW

File: api/rest.go [hunk 1]
```diff
@@ -4190,6 +4320,231 @@ func handleUpdatePublisherI18n(exec boil.Executor, id int64, i18ns []*models.Pub
 	return handleGetPublisher(exec, id)
 }
 
+//Labels
+func handleGetLabelList(exec boil.Executor, r *LabelsRequest) (*LabelsResponse, *HttpError) {
+	mods := make([]qm.QueryMod, 0)
+
+	// filters
```
Comment (edoshor) [inline api/rest.go:4327]: you must filter out label for sensitive / private units if user is not allowed to see such units.
I think inner join with content_units where secure allowed_read(cp)

File: permissions/permissions_policy.csv [hunk 2]
```diff
@@ -15,3 +15,9 @@ p, bb_user, data_public, read
 
 g, data_sensitive, data_private
 g, data_public, data_sensitive
+
+p, bb_user, data_public, label_write
+p, bb_user, data_public, label_i18n_write
+
+p, archive_label-moderation, data_public, label_read
+p, archive_label-moderation, data_public, label_moderate
```
Comment (edoshor) [inline permissions/permissions_policy.csv:23]:
```
p, bb_user, data_public, label_read
p, bb_user, data_public, label_write
p, bb_user, data_public, label_i18n_write

p, archive_label-moderation, data_sensitive, label_read
p, archive_label-moderation, data_sensitive, label_write
p, archive_label-moderation, data_sensitive, label_i18n_write
p, archive_label-moderation, data_sensitive, label_moderate
```

also need to add above

```
p, archive_editor, data_sensitive, label_read
p, archive_editor, data_sensitive, label_write
p, archive_editor, data_sensitive, label_i18n_write
```

Comment (davgur) [reply permissions/permissions_policy.csv:23]: I think something not clear for me, why need to give permission for users that dont must see it here?
1 p, bb_user, data_public, label_read - so I will need hide for them information about author (for on the future maybe will need show kc id or email)

2 why make all _data_sensitive_ for **archive_label-moderation** why not public? Its impossible to add label for cu that not public 

3 p, archive_label-moderation, data_sensitive, label_write - its impossible add label for cu from admin

4 p, archive_label-moderation, data_sensitive, label_i18n_write - you mean that usr can be **archive_label-moderation** but not **bb_user** ?

File: api/rest.go (hunk 1 above)
Comment (davgur) [reply api/rest.go:4327]: fixed

File: permissions/permissions_policy.csv (hunk 2 above)
Comment (edoshor) [reply permissions/permissions_policy.csv:23]: 1. I don't understand the problem. Try rephrasing.
2. data_sensitive includes data_public
3. For now, but api doesn't care for UI. AFAIK someone can send curl requests and it's ready for possible future with little effort.
4. No. Everybody in kc have bb_user.

'm trying to accomplish the following user stories:
* Simple user can see labels on public units
* Simple user can create new labels on public units. 
* Simple user can update the i18n of labels he created.
* archive_label-moderation user can do everything a simple user can do but also on sensitive units (not just public)
* archive_label-moderation user can moderate all labels. Sensitive and public units
* archive_editor can do everything archive_label-moderation user can do without moderation

File: api/rest.go [hunk 3]
```diff
@@ -4190,6 +4298,236 @@ func handleUpdatePublisherI18n(exec boil.Executor, id int64, i18ns []*models.Pub
...
+	}, nil
+}
+
+func handleCreateLabel(cp utils.ContextProvider, exec boil.Executor, r *CreateLabelRequest) (*models.Label, *HttpError) {
+	cu, err := models.ContentUnits(exec,
+		qm.WhereIn("uid = ?", r.ContentUnit),
+		qm.Where("secure <= ?", allowedSecure(cp, common.PERM_LABEL_WRITE)),
+	).One()
+	if err != nil {
+		return nil, NewInternalError(err)
```
Comment (edoshor) [inline api/rest.go:4362]: This 500 internal server error is confusing.
Permission errors are either 403 Forbidden or 404 Not found.

403 means: you are authenticated, I know who you are. I found the data you wish to modify. However you lack permission to do so.
404 means (in this case): I don't care who you are. I don't know what you're talking about. 

404 deliberately discloses less information

File: api/rest.go [hunk 4]
```diff
@@ -1449,6 +1448,115 @@ func StoragesHandler(c *gin.Context) {
...
+	}
+
+	var err *HttpError
+	var resp interface{}
+	if c.Request.Method == http.MethodGet || c.Request.Method == "" {
+		resp, err = handleGetLabel(c, c.MustGet("MDB").(*sql.DB), id, common.PERM_LABEL_READ)
+	}
+
+	if c.Request.Method == http.MethodPut {
+		if !can(c, secureToPermission(common.SEC_SENSITIVE), common.PERM_LABEL_MODERATE) {
```
Comment (edoshor) [inline api/rest.go:1493]: Why always checking `common.SEC_SENSITIVE` ?
This should be the secure level of the specific unit that the user want to moderate.
Some user can moderate public units and some can (in addition to that) moderate sensitive units.

File: api/rest.go [hunk 5]
```diff
@@ -4190,6 +4298,236 @@ func handleUpdatePublisherI18n(exec boil.Executor, id int64, i18ns []*models.Pub
...
+	if common.StdLang(r.I18n.Language) == common.LANG_UNKNOWN {
+		return nil, NewBadRequestError(errors.Errorf("Unknown language %s", r.I18n.Language))
+	}
+
+	label, err := models.Labels(exec,
+		qm.Where("uid = ?", uid),
+		qm.InnerJoin("content_units cu on cu.id=content_unit_id and and cu.secure <= ?", allowedSecure(cp, common.PERM_LABEL_I18N_WRITE)),
+	).One()
+	if err != nil {
+		return nil, NewInternalError(err)
```
Comment (edoshor) [inline api/rest.go:4505]: same 500 vs 40x confusion here

File: api/rest.go (hunk 4 above)
Comment (edoshor) [reply api/rest.go:1493]: I now see that both `archive_editor` and `archive_label-moderation` have sensitive access.
So no one has anything below that.
Maybe just have a comment saying that we need to change this if we change permission csv config

File: api/rest.go (hunk 3 above)
Comment (davgur) [reply api/rest.go:4362]: fixed

File: api/rest.go (hunk 4 above)
Comment (davgur) [reply api/rest.go:1493]: add comment

File: api/rest.go (hunk 5 above)
Comment (davgur) [reply api/rest.go:4505]: fixed

**Pull request merged**